syntax = "proto3";

package university;

// 1. ENUM Tanımı
enum LoanStatus {
  ONGOING = 0;   // Ödünç alma işlemi sürüyor
  RETURNED = 1;  // Kitap iade edildi
  LATE = 2;      // Gecikme var
}

// 2. Book Mesajı
message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  string isbn = 4;
  string publisher = 5;
  int32 pageCount = 6;
  int32 stock = 7;
}

// 3. Student Mesajı
message Student {
  string id = 1;
  string name = 2;
  string studentNumber = 3;
  string email = 4;
  bool isActive = 5;
}

// 4. Loan Mesajı
message Loan {
  string id = 1;
  string studentId = 2;
  string bookId = 3;
  string loanDate = 4;    // Örn: "2025-06-01"
  string returnDate = 5;  // İade edilmemişse boş string olarak kalabilir
  LoanStatus status = 6;  // LoanStatus enum’dan değer alacak
}


// 5. Request / Response Mesajları (Books)
// Listeleme
message ListBooksRequest {}
message ListBooksResponse {
  repeated Book books = 1;
}
// Tekil Görüntüleme
message GetBookRequest {
  string id = 1;
}
message GetBookResponse {
  Book book = 1;
}
// Ekleme
message CreateBookRequest {
  Book book = 1;
}
message CreateBookResponse {
  Book book = 1;
}
// Güncelleme
message UpdateBookRequest {
  Book book = 1;
}
message UpdateBookResponse {
  Book book = 1;
}
// Silme
message DeleteBookRequest {
  string id = 1;
}
message DeleteBookResponse {
  bool success = 1;
}

// 6. Request / Response Mesajları (Students)
message ListStudentsRequest {}
message ListStudentsResponse {
  repeated Student students = 1;
}

message GetStudentRequest {
  string id = 1;
}
message GetStudentResponse {
  Student student = 1;
}

message CreateStudentRequest {
  Student student = 1;
}
message CreateStudentResponse {
  Student student = 1;
}

message UpdateStudentRequest {
  Student student = 1;
}
message UpdateStudentResponse {
  Student student = 1;
}

message DeleteStudentRequest {
  string id = 1;
}
message DeleteStudentResponse {
  bool success = 1;
}

// 7. Request / Response Mesajları (Loans)
message ListLoansRequest {}
message ListLoansResponse {
  repeated Loan loans = 1;
}

message GetLoanRequest {
  string id = 1;
}
message GetLoanResponse {
  Loan loan = 1;
}

// Ödünç alma (CreateLoan)
message CreateLoanRequest {
  Loan loan = 1;
}
message CreateLoanResponse {
  Loan loan = 1;
}

// İade etme (ReturnLoan)
message ReturnLoanRequest {
  string id = 1; // ödünç kaydı ID’si
}
message ReturnLoanResponse {
  Loan loan = 1; // güncel durum (status: RETURNED, returnDate vb.)
}

// 8. Service Tanımları
service BooksService {
  rpc ListBooks (ListBooksRequest) returns (ListBooksResponse);
  rpc GetBook (GetBookRequest) returns (GetBookResponse);
  rpc CreateBook (CreateBookRequest) returns (CreateBookResponse);
  rpc UpdateBook (UpdateBookRequest) returns (UpdateBookResponse);
  rpc DeleteBook (DeleteBookRequest) returns (DeleteBookResponse);
}

service StudentsService {
  rpc ListStudents (ListStudentsRequest) returns (ListStudentsResponse);
  rpc GetStudent (GetStudentRequest) returns (GetStudentResponse);
  rpc CreateStudent (CreateStudentRequest) returns (CreateStudentResponse);
  rpc UpdateStudent (UpdateStudentRequest) returns (UpdateStudentResponse);
  rpc DeleteStudent (DeleteStudentRequest) returns (DeleteStudentResponse);
}

service LoansService {
  rpc ListLoans (ListLoansRequest) returns (ListLoansResponse);
  rpc GetLoan (GetLoanRequest) returns (GetLoanResponse);
  rpc CreateLoan (CreateLoanRequest) returns (CreateLoanResponse);
  rpc ReturnLoan (ReturnLoanRequest) returns (ReturnLoanResponse);
}
